### PR TITLE
feat: Add Area chart migration and tweaks the Timeseries chart migration

### PR DESCRIPTION
**Pull request by michael-s-molina**
### SUMMARY
This PR adds the Area chart migration logic (NVD3 ➡️ ECharts) and tweaks the base class used by timeseries migrations. Users can execute this migration using the https://github.com/apache/superset/pull/25304 CLI command and disable the legacy version with the [VIZ_TYPE_DENYLIST](https://github.com/apache/superset/blob/ec6191023263ac5f8292dd37f037805c3196e029/superset/config.py#L829) configuration.

@sadpandajoe @jinghua-qa

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Screenshot 2023-11-10 at 14 28 21
Screenshot 2023-11-10 at 14 28 46

### TESTING INSTRUCTIONS
1 - Upgrade a Area chart using the CLI command
2 - Check the new chart
3 - Downgrade a Area chart using the CLI command
4 - Check the legacy chart

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: pytest.ini [hunk 1]
```diff
@@ -17,4 +17,4 @@
 [pytest]
 testpaths =
     tests
-python_files = *_test.py test_*.py *_tests.py
+python_files = *_test.py test_*.py *_tests.py *viz/utils.py
```
Comment (michael-s-molina) [inline pytest.ini:20]: Makes `pytest` display full assertion errors defined in that file.

File: tests/unit_tests/migrations/viz/dual_line_to_mixed_chart_test.py [hunk 2]
```diff
@@ -68,30 +69,4 @@ def test_migration() -> None:
 
 
 def upgrade_downgrade(source, target) -> None:
```
Comment (john-bodley) [inline tests/unit_tests/migrations/viz/dual_line_to_mixed_chart_test.py:71]: This is legacy, but this method name is somewhat confusing, i.e., I'm not _entirely_ sure what it does.

File: tests/unit_tests/migrations/viz/dual_line_to_mixed_chart_test.py [hunk 3]
```diff
@@ -68,30 +69,4 @@ def test_migration() -> None:
...
-
-    # verify query_context
-    new_query_context = json.loads(slc.query_context)
-    assert new_query_context["form_data"]["viz_type"] == "mixed_timeseries"
-
-    # downgrade
-    slc = MigrateDualLine.downgrade_slice(slc)
-    assert slc.viz_type == MigrateDualLine.source_viz_type
-    assert json.loads(slc.params) == source
+    migrate_and_assert(MigrateDualLine, source, target)
```
Comment (john-bodley) [inline tests/unit_tests/migrations/viz/dual_line_to_mixed_chart_test.py:72]: Thanks for the refactor.

**Review by john-bodley — COMMENTED** [review]
Given there's also a refactor in combination with the are chart migration it might be helpful to annotate (via comments) which parts represent new logic.

I'm a tad complexed, given that the PR title states we're adding an area chart migration yet there was an existing "MigrateAreaChart" class.

File: tests/unit_tests/migrations/viz/dual_line_to_mixed_chart_test.py (hunk 2 above)
Comment (michael-s-molina) [reply tests/unit_tests/migrations/viz/dual_line_to_mixed_chart_test.py]: Removed 👍🏼

File: superset/migrations/shared/migrate_viz/base.py [hunk 4]
```diff
@@ -160,9 +160,7 @@ def upgrade(cls, session: Session) -> None:
         slices = session.query(Slice).filter(Slice.viz_type == cls.source_viz_type)
         for slc in paginated_update(
             slices,
-            lambda current, total: print(
-                f"  Updating {current}/{total} charts", end="\r"
-            ),
+            lambda current, total: print(f"Upgraded {current}/{total} charts"),
```
Comment (michael-s-molina) [inline superset/migrations/shared/migrate_viz/base.py:163]: Display a log entry for each block creating a log history.

**Comment by michael-s-molina** [comment]
> I'm a tad complexed, given that the PR title states we're adding an area chart migration yet there was an existing "MigrateAreaChart" class.

Even though we had a previous migration, it was incomplete and wasn't being used.

File: superset/migrations/shared/migrate_viz/processors.py [hunk 5]
```diff
@@ -36,40 +36,6 @@ def _pre_action(self) -> None:
             self.data["metric"] = self.data["metrics"][0]
 
 
-class MigrateAreaChart(MigrateViz):
```
Comment (michael-s-molina) [inline superset/migrations/shared/migrate_viz/processors.py:39]: Removed the incomplete and unused migration.

File: superset/migrations/shared/migrate_viz/processors.py [hunk 6]
```diff
@@ -137,10 +103,22 @@ class MigrateSunburst(MigrateViz):
 
 class TimeseriesChart(MigrateViz):
     has_x_axis_control = True
+    rename_keys = {
```
Comment (michael-s-molina) [inline superset/migrations/shared/migrate_viz/processors.py:106]: This is not new. Just moved from Line chart.

File: superset/migrations/shared/migrate_viz/processors.py [hunk 7]
```diff
@@ -163,23 +141,19 @@ def _pre_action(self) -> None:
...
         )
 
+        if x_ticks_layout := self.data.get("x_ticks_layout"):
+            self.data["x_ticks_layout"] = 45 if x_ticks_layout == "45°" else 0
+
 
 class MigrateLineChart(TimeseriesChart):
     source_viz_type = "line"
     target_viz_type = "echarts_timeseries_line"
-    rename_keys = {
```
Comment (michael-s-molina) [inline superset/migrations/shared/migrate_viz/processors.py:170]: Moved to Timeseries.

File: tests/unit_tests/migrations/viz/pivot_table_v1_v2_test.py [hunk 8]
```diff
@@ -14,122 +14,40 @@
...
-from tests.unit_tests.conftest import with_feature_flags
+from tests.unit_tests.migrations.viz.utils import migrate_and_assert
 
-SOURCE_FORM_DATA = {
-    "adhoc_filters": [],
+SOURCE_FORM_DATA: dict[str, Any] = {
     "any_other_key": "untouched",
     "columns": ["state"],
     "combine_metric": True,
-    "granularity_sqla": "ds",
```
Comment (michael-s-molina) [inline tests/unit_tests/migrations/viz/pivot_table_v1_v2_test.py:27]: Moved the time parts to a separate test file called `time_related_fields_test.py`

File: tests/unit_tests/migrations/viz/time_related_fields_test.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    "form_data_bak": SOURCE_FORM_DATA,
+    "granularity_sqla": "ds",
+    "rowOrder": "value_z_to_a",
+    "time_range": "100 years ago : now",
+    "viz_type": "pivot_table_v2",
+}
+
+
+@with_feature_flags(GENERIC_CHART_AXES=False)
+def test_migration_without_generic_chart_axes() -> None:
```
Comment (michael-s-molina) [inline tests/unit_tests/migrations/viz/time_related_fields_test.py:39]: All this came from the `pivot_table_v1_v2_test.py`

**Comment by michael-s-molina** [comment]
> Given there's also a refactor in combination with the are chart migration it might be helpful to annotate (via comments) which parts represent new logic.

Added comments

File: superset/charts/commands/importers/v1/utils.py [hunk 10]
```diff
@@ -75,7 +75,6 @@ def migrate_chart(config: dict[str, Any]) -> dict[str, Any]:
         if isclass(class_)
         and issubclass(class_, MigrateViz)
         and hasattr(class_, "source_viz_type")
-        and class_ != processors.MigrateAreaChart  # incomplete
```
Comment (michael-s-molina) [inline superset/charts/commands/importers/v1/utils.py:78]: Added the migration back to import function now that it's complete.

File: superset/migrations/shared/migrate_viz/processors.py [hunk 11]
```diff
@@ -189,3 +163,24 @@ def _pre_action(self) -> None:
...
+        "stack": "Stack",
+        "stream": "Stream",
+    }
+
+    def _pre_action(self) -> None:
+        super()._pre_action()
+
+        self.remove_keys.add("stacked_style")
+
+        self.data["stack"] = self.stacked_map.get(
```
Comment (john-bodley) [inline superset/migrations/shared/migrate_viz/processors.py:182]: Should the `stack` key be `None` or undefined if the resulting key isn't in `self.stacked_map`?

Comment (michael-s-molina) [reply superset/migrations/shared/migrate_viz/processors.py:182]: Yes. It can be `None` or `undefined`.

Comment (john-bodley) [reply superset/migrations/shared/migrate_viz/processors.py:182]: Do we have a preference or whether we explicitly define `None`? I guess there's likely so many inconsistencies on both the front- and back-end with regards to when/where to persist values that even if we make it undefined (and thus minimizing the payload) there's no guarantee that the field wont materialize with a value of `None` on save.

File: superset/migrations/shared/migrate_viz/processors.py [hunk 12]
```diff
@@ -137,10 +103,22 @@ class MigrateSunburst(MigrateViz):
...
+        "y_axis_label": "y_axis_title",
+        "y_axis_showminmax": "truncateYAxis",
+        "y_log_scale": "logAxis",
+    }
+    remove_keys = {"contribution", "show_brush", "show_markers"}
 
     def _pre_action(self) -> None:
         self.data["contributionMode"] = "row" if self.data.get("contribution") else None
-        self.data["zoomable"] = self.data.get("show_brush") != "no"
+        self.data["zoomable"] = self.data.get("show_brush") == "yes"
```
[the remark this answers is not in the document]
Comment (john-bodley) [reply superset/migrations/shared/migrate_viz/processors.py:121]: @michael-s-molina could we change to PR description to also include "... and add tweaks to the Timeseries chart migration".

File: superset/migrations/shared/migrate_viz/processors.py (hunk 11 above)
Comment (michael-s-molina) [reply superset/migrations/shared/migrate_viz/processors.py:182]: Exactly 😆

**Review by john-bodley — APPROVED** [review]

**Pull request merged**